### ISSUE TITLE
Correctness of KL computation

### ISSUE THREAD

**Issue by tdeboissiere** [mirceamironenco/BayesianRecurrentNN#1]
I was having a look at the KL computation for the gaussian mixture and I am not sure it is correct.

[This gist](https://gist.github.com/tdeboissiere/3f1f66ac5d790a4c5b17176f46eb0ca3) shows two versions : yours (v1), and the one I think is correct (v2).

They do give two different results.

Could you confirm ?

**Comment by mirceamironenco** [comment]
They indeed give different results, and I initially implemented your version as well. 

The issue is that when you stack the two matrices 'prior1' and 'prior2' you will have a tensor of shape 2 x dim_in x dim_out. Now when you use tf.reduce_logsumexp the tf.max will be applied on the first dimension, so it will take the max over both matrices, whereas you want the max over each individual [row, column] because it's a fully factorized gaussian, i.e. everything is independent (you want max (for example you want max(x[0,0,0], x[1,0,0])). This can be accomplished by doing the sum over 1 axis and then when stacked, the max in tf.logsumexp will be computed correctly.

Let me know if this is clear, I'll try to explain it better otherwise.

**Comment by tdeboissiere** [comment]
In v2, x is flattened.
So when I stack prior1 and prior2, I have a 2D tensor of size (2, dim_in*dim_out).

Using tf.reduce_logsumexp(prior_mix, [0]) will then correctly compute log(pi * N(x_i, mu=0, sigma1) + (1-pi) * N(x_i, mu=0, sigma2)) for each individual [row, column].

I have updated my gist to include a v3 which does not use the logsumexp trick. I obtain the same result as v2.
